File: data/src/domain/system_state.py
from datetime import datetime
from enum import StrEnum
from typing import Self

from pydantic_base import StrictBaseModel


class SystemStateStatus(StrEnum):
    NOT_RUN = "NOT_RUN"
    RUNNING = "RUNNING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


class SystemStateStage(StrEnum):
    RAW_IMPORT = "RAW_IMPORT"
    CORRECTIONS = "CORRECTIONS"
    WALLET_PROJECTION = "WALLET_PROJECTION"
    ACQUISITION_DISPOSAL = "ACQUISITION_DISPOSAL"
    TAX_COMPUTATION = "TAX_COMPUTATION"


class SystemStateError(StrictBaseModel):
    exception_type: str
    message: str
    traceback: str | None = None


class SystemState(StrictBaseModel):
    status: SystemStateStatus
    stage: SystemStateStage | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    error: SystemStateError | None = None

    @classmethod
    def not_run(cls) -> Self:
        return cls(status=SystemStateStatus.NOT_RUN)

File: ui/src/types/systemState.ts
export type SystemStateStatus = "NOT_RUN" | "RUNNING" | "COMPLETED" | "FAILED";

export type SystemStateStage =
  | "RAW_IMPORT"
  | "CORRECTIONS"
  | "WALLET_PROJECTION"
  | "ACQUISITION_DISPOSAL"
  | "TAX_COMPUTATION";

export type SystemStateError = {
  exceptionType: string;
  message: string;
  traceback: string | null;
};

export type SystemState = {
  status: SystemStateStatus;
  stage: SystemStateStage | null;
  startedAt: string | null;
  finishedAt: string | null;
  error: SystemStateError | null;
};
